### PR TITLE
Implement f"" with quotes

### PR DESCRIPTION
**Pull request by nicolasstucki**
Todo list

- [x] support basic feature to print
- [ ] type checking
- [ ] error message & positions
- [x] port tests from scalac

Requirements

- error reporting facility
- get position of trees
- get type of tree
- subtype checking
- infer implicit

Problem Found

- cannot select in macro body (`sc.parts`), need to move to implementation
- cannot interpreter correctly with explicit call `new Interpolation.Helper(sc)`.
- `Seq` works, but not `List` in the API
- doesn't support `f"${x + 1}%d"`: phase error

### CODE REVIEW

**Comment by smarter** [comment]
If this ends up implementing the full f"", we'll be able to close https://github.com/lampepfl/dotty/issues/687 :)

File: compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala [hunk 1]
```diff
@@ -138,7 +138,7 @@ class Interpreter(implicit ctx: Context) {
         env(qualifier.symbol)
 
       case SeqLiteral(elems, _) =>
-        Seq(elems.map(elem => interpretTreeImpl(elem, env)))
+        elems.map(elem => interpretTreeImpl(elem, env))
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala]: 👍

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala]: Failed tests because the check file was not updated, but this bug is fixed :)

**Comment by liufengyun** [comment]
Looking closely at the commented code that doesn't compile, it has two problems:

1. When it's safe to interpret code (`Splicer.splice`)? The current implementation (`level == 0`) seems not to be the correct condition. It boils down to a theoretical question: when it's safe to run the code of `Expr[T]`?

2. There is some bug related to level checking for inline bindings. Need to investigate more.

```Scala
val x = 5
println(ff"$x%d"              // works
println(ff"${x + 1}%d")   // doesn't work
```

**Comment by nicolasstucki** [comment]
Inline binding might be lifting computations from within a quoted inlined tree which can break phase consistency. This is what is happening with the varargs.

**Comment by nicolasstucki** [comment]
Will revisit this PR when we have proper extension methods.

**Comment by liufengyun** [comment]
@nicolasstucki Should we revive this PR?

**Comment by nicolasstucki** [comment]
What a coincidence! A student has just started today a semester project on this. https://github.com/alemannosara/f-interpolators-in-Dotty-macros

**Pull request closed without merging**